src/Symfony/Component/HttpClient/Response/HttplugPromise.php

```diff
@@ -0,0 +1,243 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient\Response;
+
+use GuzzleHttp\Promise\PromiseInterface;
+use Http\Client\Exception\NetworkException;
+use Http\Promise\Promise;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface as Psr7ResponseInterface;
+use Psr\Http\Message\StreamFactoryInterface;
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ *
+ * @internal
+ */
+final class HttplugPromise implements Promise
+{
+    private $request;
+    private $response;
+    private $client;
+    private $responseFactory;
+    private $streamFactory;
+    private $result;
+    private $state = Promise::PENDING;
+    private $callbacks = [];
+
+    private static $promisePool;
+
+    /**
+     * @param ResponseInterface|\Exception $response
+     */
+    public function __construct(RequestInterface $request, $response, HttpClientInterface $client, ResponseFactoryInterface $responseFactory, StreamFactoryInterface $streamFactory)
+    {
+        if ($response instanceof \Exception) {
+            $this->reject($response);
+
+            return;
+        }
+
+        if (!$response instanceof ResponseInterface) {
+            throw new \TypeError(sprintf('Argument 2 passed to %s() must be an instance of %s|Exception, %s given.', __METHOD__, ResponseInterface::class, \gettype($response)));
+        }
+
+        $this->request = $request;
+        $this->response = $response;
+        $this->client = $client;
+        $this->responseFactory = $responseFactory;
+        $this->streamFactory = $streamFactory;
+
+        self::$promisePool = self::$promisePool ?? new \SplObjectStorage();
+        $clientPool = self::$promisePool[$client] ?? self::$promisePool[$client] = new \SplObjectStorage();
+        $clientPool[$response] = $this;
+    }
+
+    public function cancel(): void
+    {
+        if (Promise::PENDING !== $this->state) {
+            return;
+        }
+
+        $this->callbacks = [];
+
+        if ($this->response) {
+            $this->response->cancel();
+        } elseif ($this->result instanceof self || $this->result instanceof PromiseInterface) {
+            $this->result->cancel();
+        }
+
+        $this->wait(false);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function then(callable $onFulfilled = null, callable $onRejected = null): Promise
+    {
+        if ($onFulfilled || $onRejected) {
+            $this->callbacks[] = [$onFulfilled, $onRejected];
```

Comment:
I'm quite sure this is not Promise/1+ compliant here:

```
$a = //some promise
$b = $a->then('strval')->then('trim');
$c = $a->then(some_other_callback);

// etc...
```

In your implementation, `$c->then()` is actually equivalent to `$b->then()` and gets the trimmed string value as argument instead of the original result.

Reply:
A solution (that you yourselves suggested to HTTPlug curl-client btw) might be to use `guzzlehttp/promises` in the async implementation, as it got things right.

Reply:
see https://blog.domenic.me/youre-missing-the-point-of-promises/ (this is about JS promises, but this is where the promise spec comes from). Your implementation makes the same kind of mistake than old jQuery versions did AFAIK.

Reply:
I see, thanks for the link.
I'd prefer adding no dependency as that makes using the component easier.

Reply:
then you would have to build a working promise implementation in Symfony just for that, which would duplicate part of the lib.

Reply:
OK, I gave up: `sendAsyncRequest` now requires guzzle promises :)